src/ucp/core/ucp_worker.c

```diff
@@ -2227,85 +2228,114 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
     return status;
 }
 
-static void ucp_worker_put_flush_req(ucp_request_t *req)
+static void ucp_worker_discard_uct_ep_complete(ucp_request_t *req)
 {
-    ucp_ep_h ucp_ep = req->send.ep;
+    ucp_ep_h ucp_ep     = req->send.ep;
+    ucp_worker_h worker = ucp_ep->worker;
 
     UCP_EP_ASSERT_COUNTER_DEC(&ucp_ep->discard_refcount);
-    ucp_worker_flush_ops_count_dec(ucp_ep->worker);
+
+    ucp_worker_flush_ops_count_dec(worker);
     ucp_request_complete(req, send.cb, UCS_OK, req->user_data);
     ucp_ep_remove_ref(ucp_ep);
 }
 
-static unsigned ucp_worker_discard_uct_ep_destroy_progress(void *arg)
+static void ucp_worker_discard_uct_ep_hash_delete(ucp_request_t *req)
 {
-    ucp_request_t *req  = (ucp_request_t*)arg;
     uct_ep_h uct_ep     = req->send.discard_uct_ep.uct_ep;
-    ucp_ep_h ucp_ep     = req->send.ep;
-    ucp_worker_h worker = ucp_ep->worker;
+    ucp_worker_h worker = req->send.ep->worker;
     khiter_t iter;
 
-    ucp_trace_req(req, "destroy uct_ep=%p", uct_ep);
-
-    UCS_ASYNC_BLOCK(&worker->async);
-    ucp_worker_put_flush_req(req);
-    iter = kh_get(ucp_worker_discard_uct_ep_hash,
-                  &worker->discard_uct_ep_hash, uct_ep);
+    iter = kh_get(ucp_worker_discard_uct_ep_hash, &worker->discard_uct_ep_hash,
+                  uct_ep);
     if (iter == kh_end(&worker->discard_uct_ep_hash)) {
         ucs_fatal("no %p UCT EP in the %p worker hash of discarded UCT EPs",
                   uct_ep, worker);
     }
+
+    ucs_assert(kh_value(&worker->discard_uct_ep_hash, iter) == req);
     kh_del(ucp_worker_discard_uct_ep_hash, &worker->discard_uct_ep_hash, iter);
-    UCS_ASYNC_UNBLOCK(&worker->async);
+}
+
+static unsigned ucp_worker_discard_uct_ep_destroy_progress(void *arg)
+{
+    ucp_request_t *req  = (ucp_request_t*)arg;
+    uct_ep_h uct_ep     = req->send.discard_uct_ep.uct_ep;
+    ucp_ep_h ucp_ep     = req->send.ep;
+    ucp_worker_h worker = ucp_ep->worker;
 
+    ucp_trace_req(req, "destroy uct_ep=%p", uct_ep);
     uct_ep_destroy(uct_ep);
 
+    UCS_ASYNC_BLOCK(&worker->async);
+    ucp_worker_discard_uct_ep_hash_delete(req);
+    ucp_worker_discard_uct_ep_complete(req);
+    UCS_ASYNC_UNBLOCK(&worker->async);
+
     return 1;
 }
 
-static void
-ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self)
+static UCS_F_MAYBE_UNUSED int
+ucp_worker_discard_uct_ep_check_filter(const ucs_callbackq_elem_t *elem,
+                                       void *arg)
+{
+    /* No other callbacks should be scheduled for the same UCP request
+     * simultaneously */
+    ucs_assertv_always(elem->arg != arg,
+                       "req %p: there is already scheduled operation (%s)",
+                       elem->arg, ucs_debug_get_symbol_name(elem->cb));
+    return 0;
+}
+
+static void ucp_worker_discard_uct_ep_check_no_progress(ucp_request_t *req)
+{
+#if UCS_ENABLE_ASSERT
```

Comment:
why need ifdef?

Reply:
why not? isn't it slow path?
BTW, isn't hash table already providing protection against double free?
or maybe save callback id inside `req` to check it's not scheduled twice?